libs/utils/include/utils/Allocator.h

```diff
@@ -232,9 +232,12 @@ class AtomicFreeList {
         HeadPtr currentHead = mHead.load();
         while (currentHead.offset >= 0) {
             Node* const next = storage[currentHead.offset].next.load(std::memory_order_relaxed);
-            assert(!next || next >= storage);
             newHead = { next ? int32_t(next - storage) : -1, currentHead.tag + 1 };
             if (mHead.compare_exchange_weak(currentHead, newHead)) {
+                // This assert needs to occur after we have validated that there was no race condition
```

Comment:
Very good catch. Thanks! 

I think this deserves even more commenting :-)

On line 235 we do use `next` to populate `newHead`, and now we're saying that we can't rely on the value of `next` at that point. In particular we're doing `next - storage`, what if `next` already contains user data?
This works because in that case we will loop because of the changing `tag`, and no harm is done.

Maybe just add a comment explaining that too?